lib/bolt/config/validator.rb

```diff
@@ -0,0 +1,179 @@
+# frozen_string_literal: true
+
+require 'bolt/error'
+
+# This class validates config against a schema, raising an error that includes
+# details about any invalid configuration.
+#
+module Bolt
+  class Config
+    class Validator
+      def initialize
+        @errors = []
+        @path   = []
+      end
+
+      # This is the entry method for validating data against the schema.
+      # It loops over each key-value pair in the data hash and validates
+      # the value against the relevant schema definition.
+      #
+      def validate(data, schema, location = nil)
+        data.each_pair do |key, value|
+          next unless schema.key?(key)
+
+          @path.push(key)
+          validate_value(value, schema[key])
+        ensure
+          @path.pop
+        end
+
+        raise_error(location)
+      end
+
+      # Raises a ValidationError if there are any errors. All error messages
+      # created during validation are concatenated into a single error
+      # message.
+      #
+      private def raise_error(location)
+        return unless @errors.any?
+
+        message = "Invalid configuration"
+        message += " at #{location}" if location
+        message += ":\n"
+        message += @errors.map { |error| "\s\s#{error}" }.join("\n")
+
+        raise Bolt::ValidationError, message
+      end
+
+      # Validate an individual value. This performs validation that is
+      # common to all values, including type validation. After validating
+      # the value's type, the value is passed off to an individual
+      # validation method for the value's type.
+      #
+      private def validate_value(value, definition)
+        return if plugin_reference?(value, definition)
+        return unless valid_type?(value, definition)
+
+        case value
+        when Hash
+          validate_hash(value, definition)
+        when Array
+          validate_array(value, definition)
+        when String
+          validate_string(value, definition)
+        when Numeric
+          validate_number(value, definition)
+        end
```

Comment:
This seems like it should have an else that errors, just in case none of these are matched....I don't think now we would ever reach the else, but it'd be hard to identify if anything in Psych ever changed or when upgrading Ruby and we could reach an else.

Reply:
We do reach the end if the value is a boolean. However, in that case it makes sense to fall through the case statement without a match since there's nothing to validate about booleans other than their type.

I don't think that adding an `else` adds any benefit, since any potential unknown type needs to be part of an option's types in order for validation to progress past `valid_type?`.